scheduler/src/interfaces/server/agent_interface.rs

```diff
@@ -0,0 +1,124 @@
+
+        let input = inner_req.health.unwrap();
+
+        info!("Received request from Agent: {:?}", input);
+        info!(
+            "\n  - Agent CPU usage: {}\n  - Agent memory usage: {}",
+            input.cpu_avail, input.memory_avail
+        );
+
+        let hostname = inner_req.hostname.unwrap();
+
+        info!("Received request from Agent: {:?}", hostname);
+        info!(
+            "\n  - Agent host usage: {}\n  - Agent port usage: {}",
+            hostname.host, hostname.port
+        );
+
+        // Lock the Agent Pool (to ensure thread-safe access). This is a tokio Mutex, not a standard one.
+        let mut pool = self.agent_pool.lock().await;
+
+        let id = pool.generate_unique_id();
+        let score = compute_score(input.cpu_avail, input.memory_avail);
+        let new_hostname = Hostname::new(hostname.host, hostname.port);
+
+        // Create a new Agent and add it to the Pool (it gets sorted)
+        let new_agent = PoolAgent::new(id, new_hostname, score);
+
+        // Response is the newly created Agent's ID.
+        let response = proto::RegisterAgentResponse {
+            id: new_agent.get_id(),
+        };
+
+        pool.push(new_agent);
+
+        Ok(tonic::Response::new(response))
+    }
+
+    async fn report_health_status(
+        &self,
+        request: tonic::Request<tonic::Streaming<proto::HealthStatus>>,
+    ) -> Result<tonic::Response<proto::Empty>, tonic::Status> {
+        let mut stream = request.into_inner();
+
+        while let Some(health_status) = stream.next().await {
+            let status = match health_status {
+                Ok(status) => status,
+                Err(e) => {
+                    error!("Error receiving health status: {:?}", e);
+                    return Err(tonic::Status::internal("Error receiving health status"));
+                }
+            };
+
+            let health = match status.health {
+                Some(health) => health,
+                None => {
+                    error!("Health field is missing for Agent {}", status.agent_id);
+                    continue; // Skip to the next health status if health data is missing
+                }
+            };
+
+            info!(
+                "Received health status from agent {}: CPU: {}, Memory: {}",
+                status.agent_id, health.cpu_avail, health.memory_avail
+            );
+
+            // Lock the Agent Pool (to ensure thread-safe access). This is a tokio Mutex, not a standard one.
+            let mut pool = self.agent_pool.lock().await;
+
+            // Find the Agent in the Pool
+            let agent = match pool.find_agent_mut(status.agent_id) {
+                Some(agent) => agent,
+                None => {
+                    error!("Agent ID {} not found in the Pool", status.agent_id);
+                    continue; // Skip to the next health status if the agent is not found
+                }
+            };
```

Comment:
Yes ofc, i didn't realize I wrote all of that in the loop. Sorry

Reply:
I was updating my code when I remembered why it is written as such.

Thing is, I see your reasoning: in a normal scenario, each health report stream is sent by one Agent, so all agent IDs are the same across all stream requests. But, it is possible to send a different ID in each stream, to report the health status of multiple agents (for example from another, future microservice, or simply from command-line, which is one way of testing the scheduler's functionalities)

So if I understood your concerns correctly, this is intended.

Reply:
I do not agree because this endpoint is really there to manage just want stream of agent at the time. You can have multiple opened stream/thread but each of them will treat the stream of one agent.

Reply:
Did you address this ?

Reply:
if i understand it correctly the issue here is that  a stream wil only send health status from one agent so it is not necessary to find the right agent in the pool each time we receive a health status ? is that all ?

Reply:
Yes you got it right